### DISCUSSION TITLE
Plugin Dependencies

### DISCUSSION THREAD

**Discussion by Scriptkiddi** [kimai/kimai#2457] [Questions & Answers]
Hey, 
I started working on a Plugin to allow exports to LexOffice. 
I'm currently stuck on how to include dependencies like a lib for the LexOffice Api.

I tried using composer but this leads to:
```
Cannot redeclare Composer\Autoload\includeFile()
```
I could add the dependencies with a tool like https://github.com/coenjacobs/mozart but is there a cleaner way than doing that?

**Comment by kevinpapst** [comment]
Write in your plugin documentation what has to be done, eg "composer require lexoffice/api" before/after installing your plugin.

**Reply by Scriptkiddi** [reply]
Thanks for the quick answer  :D, but I'm running NixOS, the idea being that you describe your software in packages that are build and then not writable anymore. So installing dependencies after is quite the hassle. If thats the only possible way I'm willing to do it but I was wondering if there was a way of being able to ship the dependencies so that they could be installed in the plugin folder?

**Reply by kevinpapst** [reply]
Hm, understand. I guess this is more a general dependency management question than a Kimai issue. 

Maybe there are other packages build with PHP/Symfony/Composer in NixOS package manager, where you could have a look and get some inspiration?

**Reply by Scriptkiddi** [reply]
To document how this was solved:
We added the dependencies in the main package to make them available for the plugin.

**Reply by kevinpapst** [reply]
So, what I wrote initially, correct? 
Dependency management is a problematic topic. The same problem exists for LDAP users for example, which have to execute a composer call before they can use it. Not nice, but for now the only working way.

**Reply by Scriptkiddi** [reply]
Correct.

**Comment by dreiekk** [comment]
There's the [composer-merge-plugin](https://github.com/wikimedia/composer-merge-plugin) which can include multiple `composer.json` files when running `composer update`.
Not an optimal solution, since `composer update` still has to be run after a plugin is installed.
But possibly better than editing the root `composer.json` each time.

**Reply by kevinpapst** [reply]
I know and this would have to be added to Kimai core. I am planning to test it, but I see difficulties nevertheless.
The problem is that composer.lock will be updated and very likely people will accidentaly pull new version of other dependencies. 
Or they try to update and it doesn't work, because git correctly stops as composer.lock was changed upstream and in the installation.

Brrr...  Suddenly Kimai fails ... and I get all the weird support requests.  Not something I am dreaming off^^
